### ISSUE TITLE
Merge the Gutenberg Updates into WordPress Trunk for the 5.8 release

### ISSUE THREAD

**Issue by youknowriad** [WordPress/gutenberg#31322]
This is an overview issue to track the progress of the integration of the Gutenberg Projects that area meant to land in Core for the 5.8 release before feature freeze (May 25). (See [this post](https://make.wordpress.org/core/2021/04/20/full-site-editing-go-no-go-next-steps/) for the details about the included projects)

Related WordPress Trac ticket: https://core.trac.wordpress.org/ticket/52991

- [x] **Theme.json infrastructure** https://core.trac.wordpress.org/ticket/53175
   - [x] Theme.json Global Styles infrastructure. https://github.com/WordPress/wordpress-develop/pull/1287
   - [x] Theme.json Global Settings infrastructure. https://github.com/WordPress/wordpress-develop/pull/1262
   - [x] Theme.json: update merging algorithm. https://github.com/WordPress/wordpress-develop/pull/1289
   - [x] Fully backport the layout block support. https://github.com/WordPress/wordpress-develop/pull/1295
   - [x] Add a conditional check about the "classic.css" dependency in the style loader (only load it for non theme.json themes). https://github.com/WordPress/wordpress-develop/pull/1290
   - [x] Add the supportsLayout editor setting.  
   
 - [x] **Theme blocks** (This happens automatically with package updates but we need to mark these blocks as stable in the Gutenberg repository first)
   - [x] Check PHP logic required for some of these blocks like default block context required by the Query block. https://core.trac.wordpress.org/ticket/53250
   - [x] Check the extra block patterns than need to be include to Core (Social Links, Query): https://core.trac.wordpress.org/ticket/53248
   - [ ] Enable Site Logo and related REST API changes. https://core.trac.wordpress.org/ticket/53247
   - [x] Remove Post Author.
   - [ ] Figure out the frontend.js script loading for the File block.

 - [x] **Template editing in the post editor** https://core.trac.wordpress.org/ticket/53176
   - [x] Backport the Template CPT + REST API Controller + templates fetching utilities
   - [x] Backport the `block-templates` theme support flag and setting. 
   - [x] Backport the template loader logic required to render custom block templates for classic themes (not the full logic).

 - [x] **Widget Editor** https://core.trac.wordpress.org/ticket/51506
   - [x] Backport the endpoints.
   - [x] Backport the UI.
 
 - [x] **Diverse block editor features** (Package update and related changes)
   - [x] Support for the patterns directory https://core.trac.wordpress.org/ticket/53246 
   - [x] Last packages update (Gutenberg 10.7) https://core.trac.wordpress.org/ticket/52991
   - [x] Backport the duotone block support server side.
   - [x] Enabling revisions for reusable blocks. https://core.trac.wordpress.org/ticket/53072
   - [x] Update bundled block patterns.
   - [x] Block Styles lazy loading. https://core.trac.wordpress.org/ticket/50328
 
 - [x] Check whether there  are any remaining PHP changes that doesn't fall under the items above https://github.com/WordPress/gutenberg/compare/v9.9.3...v10.7.0-rc.1 (Link to updated with each Gutenberg release until 10.7.0) 

**Note** The last package update currently in trunk corresponds to Gutenberg v10.7.0

**Comment by gziolo** [comment]
I plan to publish WordPress packages to npm with code for Gutenberg v10.5 today and open a patch in WP core that updates them.

I see several PHP changes applied for https://github.com/WordPress/gutenberg/compare/v10.4.0...v10.5.0. There is a big list of updated Block Patterns. I believe that it was already backported in https://github.com/WordPress/wordpress-develop/commit/ce50c6a982be2893bc445efde616d978c2eb3e89. I will make sure that other changes are included as well, some examples:
- block supports features got some updated
- File block has now PHP file that needs to be manually exposed

**Comment by WraithKenny** [comment]
There's a major regression in 10.5.3 that breaks User-facing Website Functionality (loading the wrong template for pages with a page template selected).

You should probably hold off merging this into core. See #31399

**Comment by gziolo** [comment]
Some updates:
- Gutenberg 10.5 changes landed to WordPress core with https://github.com/WordPress/wordpress-develop/pull/1226
- I'm looking now at the patch from @aristath that gives an option to load individual block styles for rendered blocks only https://github.com/WordPress/wordpress-develop/pull/1236
- File block contains now a script that is rendered only on the frontend, but it isn't handled in WordPress core yet. A follow-up patch is required. I will open a standalone ticket for that issue as it needs some thinking. We might need to split the webpack config to handle packages and blocks seperately.

**Comment by gziolo** [comment]
> I'm looking now at the patch from @aristath that gives an option to load individual block styles for rendered blocks only WordPress/wordpress-develop#1236

Done. This patch contains more improvements like inlining CSS styles to avoid a ton of tiny CSS files downloaded in parallel. @aristath plans to share more in the detailed dev note. Awesome work 👏🏻 👏🏻 👏🏻 👏🏻 👏🏻 

There might be a follow-up task required in the Gutenberg plugin. We are investigating it today.

That leaves JS handling for the File block on the front end. 

We should also have another updated to block editor later this week when Gutenberg 10.6 drops.

**Comment by gziolo** [comment]
Gutenberg 10.6 synced with WordPress core:
https://github.com/WordPress/wordpress-develop/pull/1257

There are some follow-up items necessary:
- Site Logo block still needs to be exposed on the server with the related logic. @aristath explores alternative handling that doesn't conflict with the unit tests in WordPress core in #31994
- Post Author block needs to be removed when 10.7 gets synced
- New  `construct_wp_query_args` function needs some sanity check and maybe more structured name. It would be great to add unit tests, cc @ntsekouras.

**Comment by gziolo** [comment]
I'm about to land renaming `construct_wp_query_args` to `build_query_vars_from_query_block` (#32041) in Gutenberg that will have to be applied to WP core as well.

@ntsekouras is planning to look at writing unit tests for `build_query_vars_from_query_block`. We also need tests for `block_has_support` and new block supports integrations added.

**Comment by youknowriad** [comment]
Core has been updated with the packages for Gutenberg 10.7.0 RC and the post author block is not removed from Core.

**Comment by youknowriad** [comment]
@ntsekouras just included the Query and Social Links block patterns to Core.

**Comment by youknowriad** [comment]
We landed the theme.json Global settings. @nosolosw and @jorgefilipecosta are focusing on the style part right now.

**Comment by youknowriad** [comment]
Theme.json infrastructure is now fully backported.

**Comment by youknowriad** [comment]
And the widgets editor is in :)

**Comment by youknowriad** [comment]
Template Mode has been merged too.

**Comment by youknowriad** [comment]
Thanks all for your work here. We got all the big features in. We have a couple smaller remaining issues. Let's now coordinate using the [5.8 Must Haves](https://github.com/WordPress/gutenberg/projects/55) board. Please take a look there and consider the items there as the highest priorities at the moment.
